### DISCUSSION TITLE
Import PCB with Python

### DISCUSSION THREAD

**Discussion by RiccSil** [thliebig/openEMS-Project#85] [Q&A Usage] [answered]
Hi all,
I am quite new to openEMS, but I would like to use it to iterate over the design of MW antennas on PCB.
My language of choice is Python and I was pleased to see that openEMS does have a Python interface, yet I could not find a way to import PCB geometry with it. I see the option is available for Matlab/Octave users with hyp2mat, so is there an equivalent for Python?
Thank you very much,
Riccardo

**Comment by thliebig** [comment]
Unfortunately the hyp2mat extension only exists in Octave/Matlab and even this one is abandoned. But lots of add-ons, extensions and tools around openEMS have evolved over time (I have no overview anymore).
Maybe you can have a look if someone has implemented some PCB to openEMS importer?
Let me/everyone know if you find something!

**Comment by biergaizi** [comment, marked as answer]
First, I believe it should still be (at least theoretically) possible to use hyp2mat. Although it cannot generate Python files, but it can generate a CSXCAD XML file, which is the 3D model used by openEMS. The generated CSXCAD model can then be manually imported to Python via [`CSXCAD.ReadFromXML()`](https://docs.openems.de/python/CSXCAD/CSXCAD.html#CSXCAD.ContinuousStructure.ReadFromXML).

Then your problem would be creating a manual mesh, ports and probes, and then simulation can be started. Meshing is the most difficult part of any openEMS simulation, it can be tricky to get right, especially for beginners (I have worked on its engine code for a potential code contribution, but even I can't say I've mastered meshing). When drawing the PCB model, make sure you use an integer grid so make meshing easier.

In addition, I'm aware of three other high-level tools that can interacts with openEMS, all are at rudimentary stages:

* jcyrax/pcbmodelgen: Convert KiCAD PCB files to models for import in openEMS 
* dlharmon/pyopenems: A Python wrapper to the OpenEMS FDTD solver with Kicad footprint generation 
* matthuszagh/pyems: High-level python interface to OpenEMS with automatic mesh generation 

The first tool can parse KiCad files and generate CSXCAD models, and even includes an automeshing algorithm. But it doesn't have explicit Python support. Although I guess you can always manually run the program and import the generated CSXCAD XML files into Python (the same can be said for hyp2mat). The next two tools attempt to add high-level object creation based on openEMS, and they can export KiCad models, but unfortunately, not importing them. I have some personal experience on pyEMS, and my impression is that it provides a clean high-level API and has the potential to be a solution in the future (currently features are lacking in all of these tools).

So for openEMS, writing some code to programmatically generate the antenna pattern is so far the preferred choice. But some antenna geometry can be quite complicated. So importing CSXCAD models from hyp2mat or pcbmodelgen would be a candidate, good luck.

I'd like to add that tools for board-level simulation is, unfortunately, currently lacking in openEMS, so I don't recommend importing the entire PCB into openEMS for simulation from the start. Instead, one should treat openEMS like a virtual research lab. You don't do experiments directly on the whole PCB, but instead, you identify problems of interests and design small and specific "test fixtures" to investigate issues in a piecewise manner. This problem is not too relevant to your planar antenna simulation, but can be very relevant to signal-integrity simulations (another piece of advice for signal-integrity simulation while we're at it: don't waste time on custom signal excitation, in particular, Heaviside or Dirac pulses are far from ideal in FDTD simulations due to numerical problems. If you want to see the time-domain response of a particular signal, use a standard wideband Gassuian excitation for simulation to calculate S-parameters, which can then be analyzed in time domain with RF circuit simulators like Qucs, scikit-rf, or even ADS).

**Reply by RiccSil** [reply]
Thank you very much for the surprisingly fast and detailed answer!
As for now, I will probably start sinking my teeth into openESM simulation with Octave and hyp2mat; that looks like the most straightforward way to begin. Once I got enough experience I'll probably try the importing of an XML file into Python and see how painful it is. If I ever manage to put together a decent workflow I will update you guys.
Thanks a gain!

**Reply by VolkerMuehlhaus** [reply]
I'm new to openEMS but worked in EM simulation support of commercial solvers for 20+ years, including FDTD solvers, so here is my 2 cents worth:

I agree to biergaizi that writing code to generate the antenna shape + PCB ground plane will be most efficient here, if the goal is to create an efficient model that delivers accurate results in reasonable time, and enables geometry tuning.  Including **all** PCB details (by importing the entire PCB data) will blow up the model without adding real value for antenna analysis or optimization. 

Of course, the animations created by samerps look amazing. But for actual antenna design procedure, I recommend to keep things simple. You don't want to slow down simulation by PCB details that have no effect on antenna performance.

**Comment by samerps** [comment]
I would like to share my approach here for simulating PCB transmission lines and antennas in openEMS with the help of Blender. For example, I was able to simulate a 60GHz mmWave reference design from Texas Instruments (TIDEP-01001). 

1. The PCB was originally designed in Altium and I was able to simply import it into KiCAD
2. Export design from KiCAD as VRML file into Blender
3. Modify, adjust and fix geometries and areas of interest in Blender, also add port location and field dump planes 
4. Export geometries as STL files into openEMS and setup simulation
5. Save simulated field data as VTK files
6. Open VTK files in Paraview and perform any necessary filter operations such as clipping and warp
7. Export Paraview scene as X3D files back into Blender
8. Use Blender’s Geometry Nodes to visualise and animate

This may seem like a long work flow of exports and imports, however once you get going it becomes manageable. 

And here is an animation of the result 

https://www.youtube.com/watch?v=wOR2y7VvlSk

Alternatively, you can use the FreeCAD-OpenEMS-Export script  https://github.com/LubomirJagos/FreeCAD-OpenEMS-Export

**Reply by RiccSil** [reply]
Yeah that looks like a nice workflow, and can't argue with the results: very cool animation :)
Thank you very much for sharing!

**Comment by LubomirJagos** [comment]
It was already mention above, but if your pcb geometry is more complicated than just simple shapes and your design is already done in some PCB editor, you can use this workflow:
    1. import PCB into KiCAD
    2. use FreeCAD with fcad_pcb plugin to import PCB into FreeCAD (https://github.com/realthunder/fcad_pcb)
    3. export solids as STL files
    4. import STL files using openEMS STL functions
    
This workflow has advantages that it's possible to correctly use also vias, and almost whole PCB modeling functions, I recommend use fcad_pcb not StepUp plugin, this import all shapes as solids not shells and they will be then handled by openEMS correctly.
